language/sygus.tex

```diff
@@ -1068,29 +1083,62 @@ \subsection{Asserting Synthesis Constraints and Assumptions}
 where $(x_1, \ldots, x_n)$ and $(y_1, \ldots, y_n)$
 are tuples of variables of sort $(\sigma_1, \ldots, \sigma_n)$ and
 $\varphi_{pre}$, $\varphi_{trans}$ and $\varphi_{post}$ are formulas.
-Given that this command is well-formed, given the above definitions,
+
+When this command is well-formed, given the above definitions,
 this command is syntax sugar for:
-\[
-\begin{array}{l}
-   \paren{\vardeclkwd\mbox{ }v_1\mbox{ }\sigma_1}\\
-   \paren{\vardeclkwd\mbox{ }v'_1\mbox{ }\sigma_1}\\
-   \ldots\\
-   \paren{\vardeclkwd\mbox{ }v_n\mbox{ }\sigma_n}\\
-   \paren{\vardeclkwd\mbox{ }v'_n\mbox{ }\sigma_n}\\
-   \paren{\constraintkwd\mbox{ }\paren{\texttt{=>}\mbox{ }
-   \paren{S_{pre}\mbox{ }v_1\mbox{ }\ldots\mbox{ }v_n}\mbox{ }
-   \paren{S\mbox{ }v_1\mbox{ }\ldots\mbox{ }v_n}}}\\
-   \paren{\constraintkwd\mbox{ }\paren{\texttt{=>}\mbox{ }
-   \paren{\texttt{and}\mbox{ }\paren{S\mbox{ }v_1\mbox{ }\ldots\mbox{ }v_n}\mbox{ }
-   \paren{S_{trans}\mbox{ }v_1\mbox{ }\ldots\mbox{ }v_n\mbox{ }v'_1\mbox{ }\ldots\mbox{ }v'_n}}\mbox{ }
-   \paren{S\mbox{ }v'_1\mbox{ }\ldots\mbox{ }v'_n}}}\\
-   \paren{\constraintkwd\mbox{ }\paren{\texttt{=>}\mbox{ }
-   \paren{S\mbox{ }v_1\mbox{ }\ldots\mbox{ }v_n}\mbox{ }
-   \paren{S_{post}\mbox{ }v_1\mbox{ }\ldots\mbox{ }v_n}}}\\
-\end{array}
-\]
+\begin{lstlisting}[language=SyGuS, numbers=none, basicstyle=\ttfamily\small]
+($\vardeclkwd$ $v_1$ $\sigma_1$)
+($\vardeclkwd$ $v'_1$ $\sigma_1$)
+$\ldots$
+($\vardeclkwd$ $v_n$ $\sigma_n$)
+($\vardeclkwd$ $v'_n$ $\sigma_n$)
+($\constraintkwd$ (=> ($S_{pre}$ $v_1$ $\ldots$ $v_n$) ($S$ $v_1$ $\ldots$ $v_n$)))
+($\constraintkwd$ (=> (and ($S$ $v_1$ $\ldots$ $v_n$) ($S_{trans}$ $v_1$ $\ldots$ $v_n$ $v'_1$ $\ldots$ $v'_n$)) ($S$ $v'_1$ $\ldots$ $v'_n$)))
+($\constraintkwd$ (=> ($S$ $v_1$ $\ldots$ $v_n$) ($S_{post}$ $v_1$ $\ldots$ $v_n$)))
+\end{lstlisting}
 where $v_1, v'_1, \ldots, v_n, v'_n$ are fresh symbols.
 
+\item \sygusinline|($\constraintchckwd$ (($x_1$ $\sigma_1$) $\ldots$ ($x_m$ $\sigma_m$)) $t_1$ $\ldots$ $t_n$ $t_{head}$)|
```

Comment:
Is there a reason to use `n+1` arguments here?
I think the syntax would be cleaner if this was `t_body t_head` instead of being n-ary.  One can always use `(and t_1 ... t_n)` or `true` for `t_body`.
This makes it clearer to parse in my opinion since you get `t_body` / `t_head` without having to explain the special cases below.

Reply:
I was using n+1 arguments to "force" a conjunction -- the shape of `tbody` is `(and t_1 ... t_n)` by construction, as opposed to syntactically allowing arbitrary `tbody` and later checking for well-formedness.

Reply:
I don't have a strong opinion on this though. I can make this change. I'm unclear on what special cases are simplified by this change though -- we still have to explain two cases for `tbody` to be well-formed: `true` and `(and t_1 ... t_n)`.

Reply:
Right, I understand the motivation. In my opinion it's a bit cleaner to do this by fixing 2 arguments. Of course this is a minor detail, but this keeps parsing simple while pushing the complications to the logic (in terms of saying that t_body should be an "and"), which is the better place for this.